org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/hcr/ScalaHotCodeReplaceManager.scala

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2015 Contributor. All rights reserved.
+ */
+package org.scalaide.debug.internal.hcr
+
+import scala.collection.JavaConverters.asScalaBufferConverter
+import scala.collection.JavaConverters.mapAsJavaMapConverter
+import scala.collection.mutable.Publisher
+
+import org.eclipse.core.resources.IResourceChangeEvent
+import org.eclipse.core.resources.IResourceChangeListener
+import org.eclipse.core.resources.ResourcesPlugin
+import org.eclipse.debug.core.DebugEvent
+import org.eclipse.debug.core.DebugPlugin
+import org.scalaide.debug.internal.BaseDebuggerActor
+import org.scalaide.debug.internal.model.ScalaDebugTarget
+import org.scalaide.debug.internal.model.ScalaDebugTarget.ReplaceClasses
+import org.scalaide.debug.internal.preferences.HotCodeReplacePreferences
+import org.scalaide.logging.HasLogger
+
+import com.sun.jdi.ReferenceType
+
+import ScalaHotCodeReplaceManager.HCRFailed
+import ScalaHotCodeReplaceManager.HCRNotSupported
+import ScalaHotCodeReplaceManager.HCRResult
+
+private[internal] object ScalaHotCodeReplaceManager {
+
+  private[hcr] sealed trait HCRResult
+  private[hcr] case class HCRNotSupported(launchName: String) extends HCRResult
+  private[hcr] case class HCRFailed(launchName: String) extends HCRResult
+
+  def create(debugTargetCompanionActor: BaseDebuggerActor): Option[ScalaHotCodeReplaceManager] = {
+    if (HotCodeReplacePreferences.hcrEnabled)
+      Some(new ScalaHotCodeReplaceManager(debugTargetCompanionActor))
+    else
+      None
+  }
+}
+
+/**
+ * Monitors changed resources and notifies debug target's companion actor, when there are some changed class files
+ * which could be replaced in VM.
+ */
+class ScalaHotCodeReplaceManager private (debugTargetCompanionActor: BaseDebuggerActor) extends IResourceChangeListener with HasLogger {
+
+  import scala.collection.JavaConverters._
+
+  private lazy val visitor = new ChangedClassFilesVisitor
+
+  private[internal] def init(): Unit = {
+    ResourcesPlugin.getWorkspace().addResourceChangeListener(this, IResourceChangeEvent.POST_BUILD)
+  }
+
+  private[internal] def dispose(): Unit = {
+    ResourcesPlugin.getWorkspace().removeResourceChangeListener(this)
+  }
+
+  override def resourceChanged(event: IResourceChangeEvent): Unit = {
+    val changedClasses = getChangedClasses(event)
+    if (changedClasses.nonEmpty)
+      debugTargetCompanionActor ! ReplaceClasses(changedClasses)
+  }
+
+  private def getChangedClasses(event: IResourceChangeEvent): List[ClassFileResource] = {
+    val delta = event.getDelta()
+    if (delta == null || event.getType() != IResourceChangeEvent.POST_BUILD)
+      Nil
+    else
+      try {
```

Comment:
You're right, but then I would log errors using the Eclipse log - not the Scala IDE log.